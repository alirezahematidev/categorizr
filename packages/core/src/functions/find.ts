import { assertion, findNode, nonUniqueTreeWarning } from "../helpers";
import { MaybeNode, TreeNode } from "$core/index";

export function find<T extends TreeNode>(tree: readonly T[], id: string): MaybeNode<T> {
  assertion(tree);

  nonUniqueTreeWarning(tree, "find");

  return findNode(tree, id);
}
