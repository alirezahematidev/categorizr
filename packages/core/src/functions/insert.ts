import { Callback, TreeNode } from "$core/types";
import { clone, exception } from "../helpers";
import { findNode } from "../helpers/internal";

function insert<T extends TreeNode>(tree: readonly T[], destination: string | null, node: T): T[];
function insert<T extends TreeNode>(tree: readonly T[], destination: string | null, node: T, callback: Callback<T>): void;
function insert<T extends TreeNode>(tree: readonly T[], destination: string | null, node: T, callback?: Callback<T>) {
  const cloneTree = clone(tree);

  if (destination === null) {
    cloneTree.push(node);

    return cloneTree;
  }

  const destNode = findNode(cloneTree, destination);

  if (!destNode) throw exception("insert", "Cannot find the destination node with the given id.");

  if (!destNode.children) destNode.children = [];

  destNode.children.push(node);

  if (callback) return void callback(cloneTree);

  return cloneTree;
}

export { insert };
