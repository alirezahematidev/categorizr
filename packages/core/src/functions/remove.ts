import { Callback, TreeNode } from "$core/index";
import { assertion, clone, exception, findNode, findParent, modifyWith, nonUniqueTreeWarning } from "../helpers";

function remove<T extends TreeNode>(tree: readonly T[], id: string): T[];
function remove<T extends TreeNode>(tree: readonly T[], id: string, callback: Callback<T>): void;
function remove<T extends TreeNode>(tree: readonly T[], id: string, callback?: Callback<T>) {
  assertion(tree);

  nonUniqueTreeWarning(tree, "remove");

  const cloneTree = clone(tree);

  const node = findNode(cloneTree, id);

  if (!node) throw exception("remove", "Cannot found the node with the given id");

  const parentNode = findParent(cloneTree, node);

  if (!parentNode) {
    modifyWith(cloneTree, node.id);

    if (callback) return void callback(cloneTree);

    return cloneTree;
  }

  if (!parentNode.children) parentNode.children = [];

  modifyWith(parentNode.children, node.id);

  if (callback) return void callback(cloneTree);

  return cloneTree;
}

export { remove };
