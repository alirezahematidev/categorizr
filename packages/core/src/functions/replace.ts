import { Callback, Replacer, TreeNode } from "$core/index";
import { clone, exception, findNode, findParent, modifyWith, nonUniqueTreeWarning } from "../helpers";

function replace<T extends TreeNode>(tree: readonly T[], target: string, replacer: Replacer<T>): T[];
function replace<T extends TreeNode>(tree: readonly T[], target: string, replacer: Replacer<T>, callback: Callback<T>): void;
function replace<T extends TreeNode>(tree: readonly T[], target: string, replacer: Replacer<T>, callback?: Callback<T>) {
  nonUniqueTreeWarning(tree, "replace");

  const cloneTree = clone(tree);

  const targetNode = findNode(cloneTree, target);

  if (!targetNode) throw exception("replace", "Cannot found the target node with the given id");

  if (!replacer.id) replacer.id = target;

  const parentNode = findParent(cloneTree, targetNode);

  if (!parentNode) {
    modifyWith(cloneTree, targetNode.id, replacer);

    if (callback) return void callback(cloneTree);

    return cloneTree;
  }

  if (!parentNode.children) parentNode.children = [];

  modifyWith(parentNode.children, targetNode.id, replacer);

  if (callback) return void callback(cloneTree);

  return cloneTree;
}

export { replace };
