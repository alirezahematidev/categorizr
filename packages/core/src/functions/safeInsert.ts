import { CallbackWithError, TreeLike, TreeNode } from "$core/index";
import { clone, exception, findNode, error, nonUniqueTreeWarning, assertion } from "../helpers";

function safeInsert<T extends TreeNode>(tree: readonly T[], destination: string | null, data: TreeLike<T>): T[];
function safeInsert<T extends TreeNode>(tree: readonly T[], destination: string | null, data: TreeLike<T>, callback: CallbackWithError<T>): void;
function safeInsert<T extends TreeNode>(tree: readonly T[], destination: string | null, data: TreeLike<T>, callback?: CallbackWithError<T>) {
  assertion(tree);

  nonUniqueTreeWarning(tree, "safeInsert");

  const cloneTree = clone(tree);

  if (destination === null) {
    cloneTree.push(...(Array.isArray(data) ? data : [data]));

    return cloneTree;
  }

  const destNode = findNode(cloneTree, destination);

  if (!destNode) {
    error("safeInsert", "Cannot found the destination node with the given id.");

    if (callback) return void callback(tree, exception("safeInsert", "Cannot found the destination node with the given id."));

    return [...tree];
  }

  if (!destNode.children) destNode.children = [];

  destNode.children.push(...(Array.isArray(data) ? data : [data]));

  if (callback) return void callback(cloneTree, undefined);

  return cloneTree;
}

export { safeInsert };
