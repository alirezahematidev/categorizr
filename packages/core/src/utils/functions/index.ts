export * from "./findNode";
export * from "./findParent";
export * from "./insertNode";
export * from "./safeInsertNode";
