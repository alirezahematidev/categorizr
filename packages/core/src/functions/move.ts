import { insert } from "./insert";
import { remove } from "./remove";
import { Callback, TreeNode } from "$core/index";
import { assertion, containsNode, exception, findNode, findParent, nonUniqueTreeWarning } from "../helpers";

function move<T extends TreeNode>(tree: readonly T[], from: string, to: string | null): T[];
function move<T extends TreeNode>(tree: readonly T[], from: string, to: string | null, callback: Callback<T>): void;
function move<T extends TreeNode>(tree: readonly T[], from: string, to: string | null, callback?: Callback<T>) {
  assertion(tree);

  nonUniqueTreeWarning(tree, "move");

  const node = findNode(tree, from);

  if (containsNode(tree, from, to)) throw exception("move", "Cannot move the node into its own descendants.");

  if (!node) throw exception("move", "Cannot found the source node with the given id");

  const parentNode = findParent(tree, node);

  if (parentNode && parentNode.id === to) {
    if (callback) return void callback(tree as T[]);

    return tree as T[];
  }

  const result = insert(remove(tree, from), to, node);

  if (callback) return void callback(result);

  return result;
}

export { move };
