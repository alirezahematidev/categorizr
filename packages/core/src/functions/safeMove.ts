import { safeInsert } from "./safeInsert";
import { safeRemove } from "./safeRemove";
import { CallbackWithError, TreeNode } from "$core/index";
import { containsNode, exception, findNode, findParent, error, nonUniqueTreeWarning, assertion } from "../helpers";

function safeMove<T extends TreeNode>(tree: readonly T[], from: string, to: string | null): T[];
function safeMove<T extends TreeNode>(tree: readonly T[], from: string, to: string | null, callback: CallbackWithError<T>): void;
function safeMove<T extends TreeNode>(tree: readonly T[], from: string, to: string | null, callback?: CallbackWithError<T>) {
  assertion(tree);

  nonUniqueTreeWarning(tree, "safeMove");

  const fromNode = findNode(tree, from);

  if (containsNode(tree, from, to)) {
    error("safeMove", "Cannot move the node into its own descendants.");

    if (callback) return void callback(tree, exception("safeMove", "Cannot move the node into its own descendants."));

    return [...tree];
  }

  if (!fromNode) {
    error("safeMove", "Cannot found the source node with the given id.");

    if (callback) return void callback(tree, exception("safeMove", "Cannot found the source node with the given id."));

    return [...tree];
  }

  const parentNode = findParent(tree, fromNode);

  if (parentNode && parentNode.id === to) {
    if (callback) return void callback(tree as T[], undefined);

    return [...tree] as T[];
  }

  const result = safeInsert(safeRemove(tree, from), to, fromNode);

  if (callback) return void callback(result, undefined);

  return result;
}

export { safeMove };
