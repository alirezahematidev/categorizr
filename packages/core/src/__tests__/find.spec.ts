import { afterEach, beforeAll, beforeEach, describe, expect, it, vi } from "vitest";
import { TreeNode } from "$core/index";
import { find } from "$core/functions";

describe("find", async () => {
  let data: TreeNode[];
  const fn = vi.fn();

  beforeAll(async () => {
    const { TREE_DATA } = await vi.importActual<{ TREE_DATA: TreeNode[] }>("$core/__mocks__");

    data = TREE_DATA;
  });

  beforeEach(() => {
    fn.mockImplementation(find);
  });

  afterEach(() => {
    fn.mockReset();
  });

  it("returns undefined for empty tree", () => {
    const emptyTree: TreeNode[] = [];

    const nodeId = "1";

    expect(fn(emptyTree, nodeId)).toBeUndefined();
  });

  it("returns correct node for existing node in the tree", async () => {
    const fn = vi.fn(find);

    const nodeId = "4";

    const expectedNode = {
      id: "4",
      name: "sub-category-2",
      children: [],
    };

    expect(fn(data, nodeId)).toStrictEqual(expectedNode);
    expect(fn(data, nodeId)).toMatchSnapshot();
  });

  it("returns correct node for existing node in the tree deeper", async () => {
    const nodeId = "5";

    const expectedNode = {
      id: "5",
      name: "sub-category-3",
      children: [],
    };

    expect(fn(data, nodeId)).toStrictEqual(expectedNode);
    expect(fn(data, nodeId)).toMatchSnapshot();
  });
});
