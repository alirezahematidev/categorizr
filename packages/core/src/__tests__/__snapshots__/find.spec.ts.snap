// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`find > returns correct node for existing node in the tree 1`] = `
{
  "children": [],
  "id": "4",
  "name": "sub-category-2",
}
`;

exports[`find > returns correct node for existing node in the tree deeper 1`] = `
{
  "children": [],
  "id": "5",
  "name": "sub-category-3",
}
`;
